### DISCUSSION TITLE
TW Icons: Not All Libraries Pad Icons The Same ...

### DISCUSSION THREAD

**Discussion by TiddlyTweeter** [morosanuae/tw-icons#2] [Ideas]
I can illustrate the issue simply ...

![moro-1-Screenshot 2021-05-05 172444](https://user-images.githubusercontent.com/28830541/117167872-f1b1f180-adc7-11eb-94f9-5b54194218fe.jpg)

Notice that the icons have _different visual heights_. The "X" and the "trash" icons are TW core icons. 
The other two aren't. _They have greater padding_.

What this means in PRACTICE is that one will need to ...

- either replace the whole core icon set with another set (with consistent padding)
- OR tweak presentation icon by icon so the SVG "viewport" zooms smaller icons to better match existing TW icons.

Thoughts?

**Comment by morosanuae** [comment]
@TiddlyTweeter This indeed can be a problem. I have face it myself in the past, but I don't know exactly what are the causes for this. We need to investigate this issue in depth especially because I'm not an expert in SVG format.
Maybe we should also establish some common standard for icons. But please remember that all the corrections can only be done in batch (automatically).
So this is a list of questions regarding the structure of SVG images:

1. What elements are important and what elements not and can be excluded or simplified?
2. What elements control the appearance of the images and how we can adapt them?
3. Can we add some useful elements to other images that lack them in order to have more customization power?

Finally, can you make an investigation on your own and present the findings here for further clarification and discussion?

**Reply by TiddlyTweeter** [reply]
> @morosanuae This indeed can be a problem. I have face it myself in the past, but I don't know exactly what are the causes for this. We need to investigate this issue in depth especially because I'm not an expert in SVG format.

The SVG format can be very confusing as its NOT an image as such. It is more like a maths puzzle :-).

> Maybe we should also establish some common standard for icons. But please remember that all the corrections can only be done in batch (automatically).

I definitely think any such adjustments should WAIT until it is _much clearer_ what to do. 

> So this is a list of questions regarding the structure of SVG images:
> 
> 1. What elements are important and what elements not and can be excluded or simplified?
> 2. What elements control the appearance of the images and how we can adapt them?

One part of that is to better understand HOW TW already styles icons. Buttons, in particular, do enforce override class.

> 3. Can we add some useful elements to other images that lack them in order to have more customization power?

I think so. Part of the issue is simply "what can be styled" _without_ having to "hack" the SVG code of individual icons. 
On that I'm not sure.

> Finally, can you make an investigation on your own and present the findings here for further clarification and discussion?

Sure. I'm slow, but I will respond in time.

**Reply by TiddlyTweeter** [reply]
> @TiddlyTweeter ... Finally, can you make an investigation on your own and present the findings here for further clarification and discussion?

Quick update ... I tried various methods for changing how the icons fill space. My conclusions SO FAR ...

1. Trying to do it through external CSS (clip-path then resize) is too cumbersome. It won't work in the context of of using library icons mixed in with TW default icons.

2. What does work to a large extent is changing the (hard-coded) **viewBox** settings. For instance ...

![Screenshot 2021-05-26 125608](https://user-images.githubusercontent.com/28830541/119648495-d2840e00-be21-11eb-9855-4f6f22450f74.jpg)

**Reply by TiddlyTweeter** [reply]
> @TiddlyTweeter  Finally, can you make an investigation on your own and present the findings here for further clarification and discussion?

Some observations arising ...

1. The default TW core icons fill the "canvas" space. Usually at least one edge of a TW icon actually touches the edge of the "canvas".

2. The Libraries in your collection differ in how they fill space. Some of them a lot.

3. They can MAYBE be brought into better "visual conformity" with TW default icons by manipulating their **viewBox** settings ...

4. ... BUT I haven't yet tested if you could change an entire library and it would work because that would assume each library is internally totally consistent.  

SO, more testing needed.

**Reply by morosanuae** [reply]
@TiddlyTweeter Very nice! Maybe we're getting close! I want you to know that I read all your posts but I can't give you an immediate answer. I need to grasp them first.

Indeed, the CSS method seems a little "complicated" and is looks like a hack not a real solution. I think we should find the root cause if there is one.

Anyway, I think that the so called "canvas" (viewBox) is just one of the problems. I see that the icons have all kinds of proportions. So my questions are: 

1. Can we enlarge the width or height of an icon (the actual drawing) only by manipulate the "viewBox" parameters or is just a static canvas? 
2. What is the connection between "viewBox", "width" and "height" parameters.
3. Are there other parameters that influence the sizing and proportions, like "scale-factor" or something similar?

I want to make some tests myself also but I don't have much time for this currently. 
Of course, If you have new findings please post them here. I'm very interested!
Thanks!

**Reply by TiddlyTweeter** [reply]
> @morosanuae 
> 1. Can we enlarge the width or height of an icon (the actual drawing) only by manipulate the "viewBox" parameters or is just a static canvas?

> 2. What is the connection between "viewBox", "width" and "height" parameters.

The width & height attributes in SVG are the **ViewPORT,**. The window onto the ViewBOX.
They can be overridden by external CSS easily. 
Since our icons are **square** it won't do much trying to change them.
They share the same proportions as the ViewBox and everything scales relative to the viewPort size.

In brief, you crop and resize (proportionally) through the ViewBox (which is really a co-ordinates system).

UPDATE: Just FYI. The first two numbers in viewBox shift the image left/right and up/down (effectively "re-bounding".  NOTE that the a number like "5" REMOVES space and a number "-5" ADDS it :-). If effect you can remove whitespace this way WHEN you also change the last two numbers. The last two numbers are "zoom" ... NOTE that to make the image BIGGER you use SMALLER numbers! :-)

> 3. Are there other parameters that influence the sizing and proportions, like "scale-factor" or something similar?

**PreseveAspectRatio=** attribute and **x=** and **y=** gives a method to scale or distort the image in the viewBox to better fit the viewPort. I haven't fully tested it. The complexity of SVG can give you brain-ache :-). I don't know all the methods for resize. But ViewBox and PreseveAspectRatio= seem like the most important, I think.

UPDATE: preseveAspectRatio seems to only have practical value when the _**proportions**_ of viewPort and viewBox DIFFER from each other (I think!)

Best
Josiah

**Reply by TiddlyTweeter** [reply]
> @morosanuae So this is a list of questions regarding the structure of SVG images:
> 
> 1. What elements are important and what elements not and can be excluded or simplified?
> 2. What elements control the appearance of the images and how we can adapt them?
> 3. Can we add some useful elements to other images that lack them in order to have more customization power?

### UPDATE: 9th June

I now looked at a sample of about 30 icons from every library. Briefly ...

**Library Differences**
- Libraries differ in the way they are constructed. Some embed `fill=` & `stroke=` attributes deep inside the svg sub-elements. Controlling these is not, as far I tested, possible via external CSS styling as they currently are. Only fill & stroke in the top `<svg ...` tag can they be overriden easily by external CSS. These libraries would need modifying if we want to gain control over their sub-elements.
- SOME of those libraries can be modified fairly easily. I will make comments & give suggestions for a few of those later.

**Regarding Re-Scaling**
- A FEW of the libraries can be rescaled to better match the size of existing TW Icons.
- BUT also note that in some cases one can increase the visual harmony with existing TW icons by changing hard-coded `stroke-width` in them.
- There is a minor issue with some of the TW Icons themselves in that they often "touch" the edges of the `viewPort`. What that means in practice is applying `stroke` to them leads to the edge of the image gets slightly cropped as the stoke falls outside the viewPort. At smaller scale this is not so visible. But at larger scale it is.
- In other words, IF we RESCALE some libraries we should still leave a small area of whitespace around them in the viewPort so that `stroke` can still be applied if wanted.

**Regarding 'Duotone' Images**
- These are a bit more tricky to handle. They don't visually match standard core TW Icons, which are Monotone images.
- In practice it may be users would use duotones best in REPLACEMENT sets where their visuals would be consistent and CSS application easier when applied to a set. 

My thoughts so far.

**Comment by TiddlyTweeter** [comment]
I'm not entirely silent on this issue. Just letting you (@morosanuae) know. Just takes time :-)

This is an update to further illustrate the issue on padding.

Playing with the libraries I found issues on padding differences BETWEEN libraries but also some issues WITHIN single libraries that do NOT have one padding system.

This shows the issue for one type of icon ...

![Screenshot 2021-05-14 105208](https://user-images.githubusercontent.com/28830541/118246623-7ffe3580-b4a2-11eb-8f7d-3291178f1026.jpg)

_Practically maybe the way to go is to establish first WHICH libraries are "padding consistent"?_

Just a comment!
TT

**Comment by morosanuae** [comment]
I realize that this is not really a "padding" issue, because the icons themselves don't have style elements inside them. The padding usually is set in the CSS outside them. I think it's a combination of the following elements: "height", "width" and "viewport" and not so easy to control/modify.  Anyway it's not a simple operation of adding, removing or adjust some parameter like CSS "padding" to normalize them. We need to investigate what can be done, or to find a way to make categories based on technical aspects like identical proportions/viewport or something.

**Reply by TiddlyTweeter** [reply]
> find a way to make categories based on technical aspects like identical proportions/viewport or something.

@morosanuae I agree with all you wrote. _Your last point seems like a good starting place?_ Basically WHICH of the icon libraries will _visually_ most easily fit in with the existing TW icons?

**Reply by TiddlyTweeter** [reply]
>  Anyway it's not a simple operation of adding, removing or adjust some parameter like CSS "padding" to normalize them.

Right! One thought I had was that maybe the CSS [clip-path property](https://www.w3schools.com/cssref/css3_pr_clip-path.asp) _might_ help IF individual libraries are consistent internally on "padding" (for want of a better term! :-). I have not tested it but it looks useful, and the coding is minimal.

UPDATE: To clarify. I meant that each library FOR integration with EXISTING TW icons maybe a local "adjustment" stylesheet might do it? 
BUT IF the end user is actually REPLACING the TW default icon set  then that issue disappears since they replacing one SVG "canvas" size with another? Meaning the "look"would be consistent without any tweaking?

**Reply by TiddlyTweeter** [reply]
I been playing around with maybe doing some kind of viewer so you can more easily see how an icon fills space. For example ...
![Screenshot 2021-05-16 075033](https://user-images.githubusercontent.com/28830541/118387393-a648e000-b61e-11eb-98a2-f7eb355877d7.jpg)

The code for that is just this ...
```
<div class="tt-test-grid">
{{$:/core/images/add-comment}}
</div>

Grid CSS code taken from: https://bansal.io/pattern-css

<style>
.tt-test-grid {margin:auto;text-align:center;}
.tt-test-grid svg {height:200px;width:200px; }
.tt-test-grid svg {background-image:radial-gradient(currentColor 1px, transparent .5px);background-size:calc(10 * .5px) calc(10 * .5px);color:brown;border:1px solid brown;}
</style>
```

**Reply by TiddlyTweeter** [reply]
I did a little more towards a SVG LAB ...

```
\define styleSVG(
 filter:'' 
 fill:'purple'
 hover-color:'pink'
 background:'transparent'
 stroke:'black' 
 stroke-width:1 
 size:'200px' 
 grid-class:""
 grid-color:""
)
<div class="tt-test-style tt-test-style-size $grid-class$">{{$:/core/images/print-button}}</div>

<div class="tt-test-style">{{$:/core/images/add-comment}}</div>

<style>
.tt-test-style {margin:auto;text-align:center;}
.tt-test-style svg {
  fill:$fill$; 
  background:$background$;
  stroke:$stroke$;
  stroke-width:$stroke-width$;
}
.tt-test-style svg:hover {
  fill:$hover-color$;
}
.tt-test-style-size svg {height:$size$;width:$size$; }
.tt-test-grid-dots svg {background-image:radial-gradient(currentColor 1px, transparent .5px);background-size:calc(10 * .5px) calc(10 * .5px);color:$grid-color$;border:1px solid $grid-color$;}
</style>
\end

<!-- Only one macro will work at a time as it creates a stylesheet that will override others -->

<<styleSVG
       filter:''
         fill:'brown' 
  hover-color:'orange'
   background:''
       stroke:'green' 
 stroke-width:'2' 
         size:'200px'
   grid-class:'tt-test-grid-dots'
   grid-color:''
>> 
```

**Reply by TiddlyTweeter** [reply]
It is still early. And I'm not really a programmer. But I can see it is NOT so difficult to create a decent test environment to look at icons in more detail ...

![Screenshot 2021-05-20 203000](https://user-images.githubusercontent.com/28830541/119030480-39c93a80-b9aa-11eb-9db5-a701853cb6c4.jpg)

**Reply by TiddlyTweeter** [reply]
### Simple Comment On Metadata. Keywords etc ...

@morosanuae wrote ... "Sadly, none of the icons sources are delivered with metadata (aliases) that will make finding icons easier."

RIGHT! I was surprised how _little_ useful metadata is available for those libraries! It makes finding the icons you need hard work.

TT

**Reply by morosanuae** [reply]
I've discovered eventually that some (few) of the libraries are indeed delivered with metadata but it's not very easy to use it. Usually it's a separate XML file. But I think this is an entire different matter and maybe it deserves it's own discussion on finding a solution for this problem.

**Reply by TiddlyTweeter** [reply]
> @morosanuae, maybe it deserves it's own discussion on finding a solution for this problem.

I'd be interested if you started such a discussion. 

There are interesting issues about how to use _"sidecar files"_. I know quite a lot about them from using Lightroom and other high end image software that use them extensively..

Just a comment.
TT

**Reply by TiddlyTweeter** [reply]
> I've discovered eventually that some (few) of the libraries are indeed delivered with metadata but it's not very easy to use it. Usually it's a separate XML file. But I think this is an entire different matter and maybe it deserves it's own discussion on finding a solution for this problem.

Further comment on this issue as it is something I know a bit about.

JPG, PNG, TIFF and other bitmap images have long had EMBEDDED metadata that is extensible. "EXIF" data is the main standard.

SVG, unfortunately, never developed, as far as I know, any method to have INTERNAL metadata standards for things like _keywords / tags_. However, in theory, they could be embedded.

I am guessing that the issue you'd face is HOW to actually convert a _sidecar XML_ of metadata such that it can be accurately added to each SVG. OR added to one of the fields of each SVG tiddler?

Just thoughts!
TT

**Comment by TiddlyTweeter** [comment]
### Update ... 13th June

@morosanuae I'll wait for your updates of the icons with faults we put in Issues before doing final tests.

I'm continuing to explore the icons and possibilities. 
It is interesting. 

Once you can have time to update I'll comment in more detail with examples.

Best wishes
@TiddlyTweeter

**Comment by TiddlyTweeter** [comment]
@morosanuae, just FYI ... a few more people in this discussion would be good ... so I ask periodically for that ... https://groups.google.com/g/tiddlywiki/c/eccIEHZoxsI/m/MGVQy9LHAgAJ  - I'm not sure it will ever happen (the number of real active users is actually quite low on GG) but I do think it is worth mentioning your library when relevant. 
TW is more than JS / HTML / CSS ... it is also **SVG** :-)
TT

**Reply by morosanuae** [reply]
@TiddlyTweeter My apologies for not being too active on this. I've been very busy lately. To be honest I was counting on a larger support base for this because I don't really have time to work on this. It's not a simple 1-2 clicks matter. Maybe you're the only user who wants to use the SVGs like a pro :)

**Reply by TiddlyTweeter** [reply]
> @morosanuae  ... To be honest I was counting on a larger support base for this because I don't really have time to work on this. It's not a simple 1-2 clicks matter. 

Right!

**_I have a _Question_ for you :-) ..._**

I could, now, with some confidence, convert SOME of the libraries to better CONFORM with standard TiddlyWiki styling & sizing. BUT, I am also aware of this you wrote ...

> @morosanuae  After the initial creation of this site, my work will consist from now on mainly of updating/adding content:
Update current libraries to their latest versions.

The PROBLEM is that, because all the libraries are inconsistent, in one way or another, it is NOT possible to do this in a way that could be backwards compatible via any future bulk script re-conversions  (I think).

So, I'm happy to update some libraries, but I'm concerned it will just _make your intended later updating too difficult???_

Your thoughts?
Best wishes, TT

**Reply by morosanuae** [reply]
@TiddlyTweeter Please, make some examples that will ease and speed up the process of my (and others) understanding. You can create an example site on the new TiddlyHost service. And then I'll be able to tell you what can be done. Maybe I'll even will post an example script that can be adjusted to your needs. We'll see, but first make some conversion demos, at least for one library to have a starting point.

**Reply by morosanuae** [reply]
So we have to aspects here that should be treated individually first:
1. The modifications that are needed for conformity with TW original SVGs
2. The process of integrating those modifications in further libraries updates.

**Reply by TiddlyTweeter** [reply]
Ciao @morosanuae, you might like to know that my ONE area of competence is regular expressions which I been interested in  forever. I also understand the basics of PowerShell. I know more about them than I do about TW :-) Lol!

**Reply by TiddlyTweeter** [reply]
> So we have to aspects here that should be treated individually first:
> 
> 1. The modifications that are needed for conformity with TW original SVGs
> 2. The process of integrating those modifications in further libraries updates.

Regarding (1) the major issues that CAN be addressed are (a) SIZING of the ViewBox and (b) removal of HARD-CODE colours that external CSS can't reach.

Regarding (2). Right! You need some illustrations to show if it is feasible. Over the next week I'll do a couple of libraries as a start.

**Reply by morosanuae** [reply]
@TiddlyTweeter I'm at a very basic level in regular expressions and also in PowerShell. I've made the scripts in a couple of days without prior knowledge (only medium programming experience and a little experience with CMD). I usually make my own research only when I need something new. That's it. Nothing fancy here. I'm aware that the scripts that I made can be improved a lot, but that again as I said I don't really have time to do that.

**Comment by TiddlyTweeter** [comment]
Ciao @morosanuae ... I haven't forgotten us. Just bizzie atm. I do try mention your work when it comes up ... [TiddlyTalk](https://talk.tiddlywiki.org/t/a-nice-trick-to-use-svg-with-css-better-solution-than-font-awesome-imo/802/3?u=tiddlytweeter)

**Comment by morosanuae** [comment]
Ciao @TiddlyTweeter. I understand because I'm also way too busy at the moment to even think about this project. I've not abandoned it but, to be honest, I don't really know when I will have the time to pick it up again. Maybe this winter, I hope... Sorry!

**Reply by TiddlyTweeter** [reply]
@morosanuae FYI ... this https://talk.tiddlywiki.org/t/recommand-a-icon-search-tool-to-your-guys/1498/2?u=tiddlytweeter and this https://talk.tiddlywiki.org/t/how-to-set-the-favicon-of-website/1309/8?u=tiddlytweeter

**Comment by TiddlyTweeter** [comment]
@morosanuae, a quick update post that may interest you: [[Easiest method of editing Icons?](https://talk.tiddlywiki.org/t/easiest-method-of-editting-icons/3892)](https://talk.tiddlywiki.org/t/easiest-method-of-editting-icons/3892?u=tiddlytweeter )
